db/commitlog/commitlog.cc

```diff
@@ -924,133 +923,134 @@ class db::commitlog::segment : public enable_shared_from_this<segment>, public c
         });
     }
 
+    void background_cycle() {
+        //FIXME: discarded future
+        (void)cycle().discard_result().handle_exception([] (auto ex) {
+            clogger.error("Failed to flush commits to disk: {}", ex);
+        });
+    }
+
     /**
      * Add a "mutation" to the segment.
+     * Should only be called from "allocate_when_possible". "this" must be secure in a shared_ptr that will not
+     * die. We don't keep ourselves alive (anymore)
      */
     future<> allocate(shared_ptr<entry_writer> writer, segment_manager::request_controller_units permit, db::timeout_clock::time_point timeout) {
-        if (must_sync()) {
-            return with_timeout(timeout, sync()).then([this, writer = std::move(writer), permit = std::move(permit), timeout] (auto s) mutable {
-                return s->allocate(std::move(writer), std::move(permit), timeout);
-            });
-        }
+        for (;;) {
+            if (must_sync()) {
+                co_await with_timeout(timeout, sync());
+                continue;
```

Comment:
This is `while (must sync) co_await sync()`. The `continue` is especially odd. We can clean it up later.

I expect this generates a small regression in the happy case that memory is available, since we'll allocate the coroutine frame and immediately delete it. Please follow up splitting this into two - a normal function to detect and handle the happy case, and a coroutine for all the complexity.

Reply:
My understanding is that (at least) the whole function becomes a coroutine frame, and we just iterate _inside_ same allocation. Opted for not inner loop due to the pattern that in certain cases we want to jump to the top of the routine (see below), and I want them all symmetric.

Reply:
Changed in refactored version.